client/client.go

```diff
@@ -2,24 +2,90 @@ package client
 
 import (
 	"context"
+	"encoding/json"
 	"fmt"
 	"io"
 	"net/http"
+	"time"
 
 	"github.com/deepmap/oapi-codegen/pkg/securityprovider"
 	"github.com/hashicorp/go-cleanhttp"
 	"github.com/hashicorp/go-retryablehttp"
 	"github.com/pkg/errors"
+	"golang.org/x/time/rate"
 )
 
+// last retry will wait 64 seconds (2^6)
+// will also listen to 'Retry-After' header
+
+const (
+	rpm              = 600
+	bucketsPerMinute = 60
+	maxRetries       = 6
+	minRetryWait     = 1 * time.Second
+	maxRetryWait     = 65 * time.Second
+)
+
+type RateLimitedClient struct {
+	client      *http.Client
+	rateLimiter *rate.Limiter
+}
+
+func (c *RateLimitedClient) Do(req *http.Request) (*http.Response, error) {
+	if c.rateLimiter == nil {
+		return nil, errors.New("rate limiter not set")
+	}
+	ctx := req.Context()
+	err := c.rateLimiter.Wait(ctx)
+	if err != nil {
+		return nil, err
+	}
+	resp, err := c.client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	return resp, nil
+}
+
+func attentiveBackoff(min, max time.Duration, attemptNum int, resp *http.Response) time.Duration {
+	// Retry for rate limits and server errors.
+	if resp != nil && resp.StatusCode == http.StatusTooManyRequests || resp.StatusCode >= 500 {
```

Comment:
perhaps wrong but we only expect a rate limit response on a 429 right? I'd expect 500 to fall through to the default backoff

Reply:
absolutely